src/pytesmo/validation_framework/metric_calculators_adapters.py

```diff
@@ -24,73 +24,297 @@
-    PairwiseIntercomparisonMetrics,
-    TripleCollocationMetrics
-)
+    PairwiseIntercomparisonMetrics, TripleCollocationMetrics)
 import warnings
 import numpy as np
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Union, List, Tuple, Optional
+from cadati.conv_doy import doy, days_past
+import calendar
 
 
-class MonthsMetricsAdapter(object):
+def days_in_month(month: int) -> int:
     """
-    Adapt MetricCalculators to calculate metrics for groups across months
+    Get number of days in this month (in a LEAP YEAR)
+    """
+    return days_past[month] - days_past[month - 1]
+
+
+@dataclass
+class GenericDatetime:
+    """
+    Datetime without a year. Only down to second.
+    """
+    month: int
+
+    day: int = 1
+    hour: int = 0
+    minute: int = 0
+    second: int = 0
+
+    @property
+    def __ly(self):
+        return 2400  # arbitrary leap year
+
+    def __ge__(self, other: 'GenericDatetime'):
+        return self.to_datetime(self.__ly) >= other.to_datetime(self.__ly)
+
+    def __le__(self, other: 'GenericDatetime'):
+        return self.to_datetime(self.__ly) <= other.to_datetime(self.__ly)
+
+    def __lt__(self, other: 'GenericDatetime'):
+        return self.to_datetime(self.__ly) < other.to_datetime(self.__ly)
+
+    def __gt__(self, other: 'GenericDatetime'):
+        return self.to_datetime(self.__ly) > other.to_datetime(self.__ly)
+
+    def __repr__(self):
+        return f"****-{self.month:02}-{self.day:02}" \
+               f"T{self.hour:02}:{self.minute:02}:{self.second:02}"
+
+    @property
+    def doy(self) -> int:
+        """
+        Get day of year for this date. Assume leap year!
+        i.e.: 1=Jan.1st, 366=Dec.31st, 60=Feb.29th.
+        """
+        return doy(self.month, self.day, year=None)
+
+    @classmethod
+    def from_datetime(cls, dt: datetime):
+        """
+        Omit year from passed datetime to create generic datetime.
+        """
+        return cls(dt.month, dt.day, dt.hour, dt.minute, dt.second)
+
+    def to_datetime(self, years: Optional[Union[Tuple[int, ...], int]]) \
+            -> Union[datetime, List, None]:
+        """
+        Convert generic datetime to datetime with year.
+        Feb 29th for non-leap-years will return None
+        """
+        dt = []
+
+        for year in np.atleast_1d(years):
+            if not calendar.isleap(year) and self.doy == 60.:
+                continue
+            else:
+                d = datetime(year, self.month, self.day, self.hour,
+                             self.minute, self.second)
+            dt.append(d)
+
+        if len(dt) == 1:
+            return dt[0]
+        elif len(dt) == 0:
+            return None
+        else:
+            return dt
+
+
+class TsDistributor:
+
+    def __init__(self,
+                 dates=None,
+                 date_ranges=None,
+                 generic_dates=None,
+                 generic_date_ranges=None):
```

Comment:
I'm not a fan of long keyword argument names, they are always hard to remember. I would instead use
1) 2 keywords, `dates` and `date_ranges`, and do type/attribute checks to figure out whether it's a datetime or a GenericDatetime
2) or accept only a single list of individual datetimes, GenericDatetimes, or tuples of datimes and GenericDatetimes.

As a more general comment about design:

From the tests and how you used the class in the code I also got the impresssion that only the `select` method of the class is really supposed to be used. Since the class doesn't really have a state it might be conceptually simpler to make a function out of this (e.g. select_subset).

If we implement option 2 from above, we could also simplify the interface of SubsetsMetricAdapter to accept a dictionary of lists instead of a dictionary of TsDistributors, and make SubsetsMetricAdapter generate the TsDistributors/call select_subset using the lists as arguments. This would make SubsetsMetricAdapter a bit easier to use.

Reply:
Okay, then we just have different preferences on this :D You can also leave it then, just an idea I had when going through the code.